File: server/services/nbaHints.ts
import fs from 'fs';
import path from 'path';
import { CharacterEntry, HintInfo, THEME_FIELDS } from '../types';

const { divisions, positions } = (() => {
  try {
    const raw = fs.readFileSync(path.join(__dirname, '..', 'data', 'nba.json'), 'utf-8');
    const parsed = JSON.parse(raw);
    if (Array.isArray(parsed)) return { divisions: {}, positions: {} };
    return {
      divisions: (parsed.meta?.divisions as Record<string, string>) ?? {},
      positions: (parsed.meta?.positions as Record<string, string>) ?? {},
    };
  } catch {
    return { divisions: {}, positions: {} };
  }
})();

const playableMinTotalGpSince2025 = (() => {
  try {
    const raw = fs.readFileSync(path.join(__dirname, '..', 'data', 'nba.json'), 'utf-8');
    const parsed = JSON.parse(raw);
    const v = parsed.meta?.playableMinTotalGpSince2025;
    return typeof v === 'number' && v >= 0 ? v : 0;
  } catch {
    return 0;
  }
})();

export function getPlayableMinTotalGpSince2025(): number {
  return playableMinTotalGpSince2025;
}

const HINT_EXCLUDE = new Set([
  'name',
  'id',
  'aliases',
  'displayName',
  'imageUrl',
  'englishName',
]);

export function getDivisionHint(team: unknown): string | null {
  const code = String(team || '').trim();
  if (!code) return null;
  return divisions[code] || null;
}

function getPositionZh(pos: unknown): string {
  const code = String(pos || '').trim();
  if (!code) return '未知位置';
  return positions[code] || code;
}

const DRAFT_ROUND_LABELS: Record<number, string> = {
  1: '首轮秀',
  2: '次轮秀',
};

/** 选秀字符串中仅保留轮次展示（不含年份），如「2024年第2轮」→「次轮秀」 */
export function getDraftRoundHint(draft: unknown): string | null {
  const s = String(draft || '').trim();
  if (!s) return null;
  if (s === '落选秀') return '落选秀';
  const m = s.match(/第(\d+)轮/);
  if (m) {
    const round = parseInt(m[1], 10);
    return DRAFT_ROUND_LABELS[round] ?? `第${round}轮`;
  }
  const stripped = s.replace(/^\d{4}年/, '').trim();
  return stripped || null;
}

/** Compare fields usable as extra hints (excludes division composite field key). */
export function getNBAExtraHintFields(): string[] {
  return (
    THEME_FIELDS.nba?.map((f) => f.field).filter(
      (k) => !HINT_EXCLUDE.has(k) && k !== 'division' && k !== 'divisionPosition'
    ) ?? []
  );
}

/** First NBA hint: division + draft round (year omitted). */
export function buildNBAPrimaryHint(answer: CharacterEntry): HintInfo {
  const division = getDivisionHint(answer.team);
  const divLabel = division ? division.replace(/球员$/, '') : '未知赛区';
  const roundLabel = getDraftRoundHint(answer.draft) ?? '未知轮次';
  return {
    field: 'divisionPosition',
    label: '赛区·选秀轮次',
    value: `${divLabel} · ${roundLabel}`,
  };
}

export function isNBAHintField(field: string): boolean {
  return field === 'division' || field === 'divisionPosition';
}

export function isNBAHintFieldExcluded(field: string): boolean {
  return (
    HINT_EXCLUDE.has(field) ||
    field === 'division' ||
    field === 'divisionPosition' ||
    field === 'position'
  );
}

/** Hupu 2025+ career data + min games since 2025 (meta.playableMinTotalGpSince2025). */
export function isPlayableNBAAnswer(entry: CharacterEntry): boolean {
  if (entry.hasCareerSince2025 !== true) return false;
  const minGp = playableMinTotalGpSince2025;
  if (minGp <= 0) return true;
  const total = entry.totalGpSince2025;
  const best = entry.bestGpSince2025;
  const gp =
    typeof total === 'number' && typeof best === 'number'
      ? Math.max(total, best)
      : typeof best === 'number'
        ? best
        : total;
  if (gp == null || typeof gp !== 'number') return false;
  return gp >= minGp;
}

export function getNBAPositionLabel(entry: CharacterEntry): string {
  return getPositionZh(entry.position);
}

File: server-go/internal/services/nba_hints.go
package services

import (
	"fmt"
	"regexp"
	"strings"

	"guess-who/server-go/internal/types"
)

var draftRoundLabels = map[int]string{
	1: "首轮秀",
	2: "次轮秀",
}

var nbaDraftRoundRe = regexp.MustCompile(`第(\d+)轮`)

func loadNBAConfig() (divisions, positions, teams map[string]string, minGp int) {
	doc, err := LoadTheme(types.ThemeNBA)
	if err != nil {
		return nil, nil, nil, 0
	}
	divisions = mapFromConfig(doc.Config, "divisions")
	positions = mapFromConfig(doc.Config, "positions")
	teams = mapFromConfig(doc.Config, "teams")
	if v, ok := doc.Config["playableMinTotalGpSince2025"].(float64); ok && v >= 0 {
		minGp = int(v)
	}
	return divisions, positions, teams, minGp
}

func GetDivisionHint(team string, divisions map[string]string) string {
	code := strings.TrimSpace(team)
	if code == "" || divisions == nil {
		return ""
	}
	if v, ok := divisions[code]; ok {
		return v
	}
	return ""
}

func GetDraftRoundHint(draft string) string {
	s := strings.TrimSpace(draft)
	if s == "" {
		return ""
	}
	if s == "落选秀" {
		return "落选秀"
	}
	m := nbaDraftRoundRe.FindStringSubmatch(s)
	if len(m) == 2 {
		round := 0
		fmt.Sscanf(m[1], "%d", &round)
		if label, ok := draftRoundLabels[round]; ok {
			return label
		}
		return fmt.Sprintf("第%d轮", round)
	}
	return ""
}

func GetNBAExtraHintFields() []string {
	excluded := map[string]bool{
		"name": true, "id": true, "aliases": true, "displayName": true,
		"imageUrl": true, "englishName": true, "division": true, "divisionPosition": true,
	}
	var out []string
	for _, def := range types.ThemeFieldDefs[types.ThemeNBA] {
		if !excluded[def.Field] {
			out = append(out, def.Field)
		}
	}
	return out
}

func BuildNBAPrimaryHint(answer types.CharacterEntry) types.HintInfo {
	divisions, _, _, _ := loadNBAConfig()
	division := GetDivisionHint(fmt.Sprintf("%v", answer["team"]), divisions)
	divLabel := "未知赛区"
	if division != "" {
		divLabel = strings.ReplaceAll(division, "球员", "")
	}
	roundLabel := GetDraftRoundHint(fmt.Sprintf("%v", answer["draft"]))
	if roundLabel == "" {
		roundLabel = "未知轮次"
	}
	return types.HintInfo{
		Field: "divisionPosition",
		Label: "赛区·选秀轮次",
		Value: fmt.Sprintf("%s · %s", divLabel, roundLabel),
	}
}

func IsNBAHintField(field string) bool {
	return field == "division" || field == "divisionPosition"
}

func IsNBAHintFieldExcluded(field string) bool {
	excluded := map[string]bool{
		"name": true, "id": true, "aliases": true, "displayName": true,
		"imageUrl": true, "englishName": true, "division": true,
		"divisionPosition": true, "position": true,
	}
	return excluded[field]
}

func GetNBAPositionLabel(entry types.CharacterEntry) string {
	_, positions, _, _ := loadNBAConfig()
	code := strings.TrimSpace(fmt.Sprintf("%v", entry["position"]))
	if code == "" {
		return "未知位置"
	}
	if positions != nil {
		if p, ok := positions[code]; ok {
			return p
		}
	}
	return code
}

func GetDivisionPosition(team, position string, divisions, positions map[string]string) string {
	div := GetDivisionHint(team, divisions)
	pos := position
	if positions != nil {
		if p, ok := positions[position]; ok {
			pos = p
		}
	}
	if div != "" && pos != "" {
		return fmt.Sprintf("%s-%s", div, pos)
	}
	return ""
}

func IsPlayableNBA(entry types.CharacterEntry, minGp int) bool {
	if entry.Get("hasCareerSince2025") != true {
		return false
	}
	_, _, _, cfgMin := loadNBAConfig()
	if minGp <= 0 {
		minGp = cfgMin
	}
	if minGp <= 0 {
		return true
	}
	total, _ := entry["totalGpSince2025"].(float64)
	best, _ := entry["bestGpSince2025"].(float64)
	gp := total
	if best > gp {
		gp = best
	}
	return gp >= float64(minGp)
}
